rust/src/transformations/clamp/make_clamp.tex

```diff
@@ -1,65 +1,65 @@
-
-    For the second precondition, assume the input is a member of \texttt{input\_domain}'s row domain. 
-    Therefore, by \ref{line:assert-non-null}, the input is non-null.
-    In addition, since \texttt{Bounds.new\_closed} did not raise an exception, then by the definition of \texttt{Bounds.new\_closed}, the bounds are non-null.
-    Thus, by the definition of \rustdoc{traits/trait}{ProductOrd}, the preconditions of \texttt{total\_clamp} are satisfied, so the output is within the bounds.
-    Therefore, the output is a member of \texttt{output\_row\_domain}.
-\end{proof}
-
-We now prove the postcondition of \texttt{make\_clamp}.
-\begin{proof}
-By \ref{lemma:row-by-row-precondition}, the preconditions of \rustdoc{transformations/manipulation/fn}{make\_row\_by\_row\_fallible} are satisfied.
-Thus, by the definition of \\\texttt{make\_row\_by\_row\_fallible}, the output is a valid transformation.
-\end{proof}
-
+\documentclass{article} 
+\input{../../lib.sty} 
+ 
+\title{\texttt{fn make\_clamp}} 
+\author{S\'ilvia Casacuberta} 
+\date{} 
+ 
+\begin{document} 
+ 
+\maketitle 
+ 
+Proves soundness of \texttt{fn make\_clamp} in \asOfCommit{mod.rs}{0db9c6036}. 
+ 
+\section{Hoare Triple} 
+\subsection*{Precondition} 
+ 
+\textbf{Compiler-verified} 
+\begin{itemize} 
+    \item Type \texttt{TA} must have traits \texttt{Clone + ProductOrd + CheckAtom}. 
+    \item Type \texttt{M} must have trait \texttt{DatasetMetric}. 
+    \item Argument \texttt{input\_domain} is of type \texttt{VectorDomain<AtomDomain<TA>>}. 
+    \item Argument \texttt{input\_metric} is of type \texttt{M}. 
+    \item Argument \texttt{bounds} is of type \texttt{(T, T)}. 
+\end{itemize} 
+ 
+\textbf{User-verified} 
+None 
+ 
+\subsection*{Pseudocode} 
+\lstinputlisting[language=Python,firstline=2,escapechar=|]{./pseudocode/make_clamp.py} 
+ 
+\subsubsection*{Postconditions} 
+\validTransformation{\texttt{(input\_domain, input\_metric, bounds, TA, M)}}{\texttt{make\_clamp}} 
+ 
+\section{Proof} 
+ 
+\begin{lemma} 
+    The invocation of \rustdoc{transformations/manipulation/fn}{make\_row\_by\_row\_fallible} (line \ref{line:row-by-row}) satisfies its preconditions. 
+\end{lemma} 
+ 
+\begin{proof} 
+    \label{lemma:row-by-row-precondition} 
+    The preconditions of \texttt{make\_clamp} and pseudocode definition (line \ref{line:def}) ensure that the type preconditions of \texttt{make\_row\_by\_row\_fallible} are satisfied.  
+    The remaining preconditions of \texttt{make\_row\_by\_row\_fallible} are: 
+    \begin{itemize} 
+        \item \texttt{row\_function} has no side-effects. 
+        \item If the input to \texttt{row\_function} is a member of \texttt{input\_domain}'s row domain, then the output is a member of \texttt{output\_row\_domain}. 
+    \end{itemize} 
+ 
+    The first precondition is satisfied by the definition of \texttt{clamper} (line \ref{line:clamper}) in the pseudocode. 
```

Comment:
The pseudocode calls a function `total_clamp` which is not defined anywhere in the proof document (but is perhaps a function defined by one of the traits?). It is not clear to me that this function does not have side-effects without knowing what this function is doing.

Reply:
I added a comment and doc link, and updated the proof definition.